feat: working instrumentation with smolagents

aymeric-roucher opened the pull request:
Fixes https://github.com/Arize-ai/openinference/issues/1182
cc @harrisonchu I've copied the work you did for crewAI, ended up working really well!

I've put a `test.py` file at the root to let you try out the instrumentation.

Beware that I've not adapted the tests yet.

axiomofjoy commented:
Excited to dig into this @aymeric-roucher!

aymeric-roucher commented on python/instrumentation/openinference-instrumentation-smolagents/src/openinference/instrumentation/smolagents/_wrappers.py:
```diff
@@ -0,0 +1,270 @@
+                    agent.model.last_input_token_count + agent.model.last_output_token_count
+                )
+                span.set_attribute("Observations", step_log.observations)
+                # if step_log.error is not None:
```
@harrisonchu As you can see I've commented out these 3 lines.
It is because currently their visualization in a platform like Arize Phoenix is unsatisfactory: having an error in one step shows the whole run as failing, when indeed I can have an error at one step but then the multi-step agent recovers in the next steps to successfully solve the task. Is there a way to display an error without upwards transmission of the error to the whole run?

![image](https://github.com/user-attachments/assets/2196afc4-70ca-43b8-9f21-413d81529049)

axiomofjoy replied:
Interesting, can you give me a code snippet to reproduce this behavior?

aymeric-roucher replied:
This is the code in script test.py

axiomofjoy replied:
Will take a look and get back to you!

axiomofjoy replied:
I seem to be hitting this issue running the script.

```
Traceback (most recent call last):
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/opentelemetry/sdk/trace/export/__init__.py", line 113, in on_end
    self.span_exporter.export((span,))
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py", line 189, in export
    return self._export_serialized_spans(serialized_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py", line 159, in _export_serialized_spans
    resp = self._export(serialized_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py", line 133, in _export
    return self._session.post(
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/requests/sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/.venv/lib/python3.11/site-packages/requests/adapters.py", line 700, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='0.0.0.0', port=6006): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x13c979ed0>: Failed to establish a new connection: [Errno 61] Connection refused'))
Traceback (most recent call last):
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/test.py", line 34, in <module>
    manager_agent.run(
  File "/Users/xandersong/openinference/python/instrumentation/openinference-instrumentation-smolagents/src/openinference/instrumentation/smolagents/_wrappers.py", line 126, in __call__
    span.set_attribute("tools_names", list(agent.tools.keys()))
                                           ^^^^^^^^^^^
AttributeError: 'CodeAgent' object has no attribute 'tools'
```

axiomofjoy replied:
I think this might be a bug in Phoenix. I have filed an issue [here](https://github.com/Arize-ai/phoenix/issues/5981).

aymeric-roucher commented:
Nice seeing you here @axiomofjoy! 🤗

axiomofjoy commented:
Great to meet you @aymeric-roucher and thanks so much for this contribution! I'm digging into the PR now. Let me know how I can help you get it over the line, e.g., if you need help writing tests. Our team would also love to dogfood the instrumentation once it's in a good state!

aymeric-roucher commented:
@axiomofjoy sorry for the issue above, I had been working on a specific branch of smolagents to make it compatible: now installing from main should work!

aymeric-roucher commented:
@axiomofjoy one more problem that I just detected: detecting calls to [Model.get_tool_call](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/models.py#L160) from within [ToolCallingAgent](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/agents.py#L779) does not work. I think this is because I actually use [HfApiModel](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/models.py#L274) as my LLM: `HfApiModel` inherits from `Model` but since it overrides its parent's `get_tool_call` method, maybe the wrapper targeting `Model.get_tool_call()` does not work for `HfApiModel`.

Is there a way to create a wrapper that works for `Subclass.get_tool_call` for any subclass of `Model`?

Method `__call__` works for standard LLM calls (LLM calls that are just "please generate some text", not "please generate a tool call") because it's defined in Model and not overridden. But it would seem like a hack to build such a wrapper method for `Model.get_tool_call`, so I'd prefer a "wrapper that works for this classe's method `get_tool_call` and all methods from child classes with name `get_tool_call`"


![image](https://github.com/user-attachments/assets/4acccc83-fdb5-40c2-808e-beaca316a6c9)

aymeric-roucher commented:
I have read the CLA Document and I hereby sign the CLA

aymeric-roucher commented:
Also wondering how to make the nice LLM messages format with system/user shown in this screenshot:
![image](https://github.com/user-attachments/assets/8250f4bd-ab21-4286-b304-2acbab4c3ccf)

axiomofjoy commented:
> @axiomofjoy sorry for the issue above, I had been working on a specific branch of smolagents to make it compatible: now installing from main should work!

Got it, no worries! I just got it running from dev 😄

axiomofjoy commented:
> Also wondering how to make the nice LLM messages format with system/user shown in this screenshot: ![image](https://private-user-images.githubusercontent.com/69208727/401715758-8250f4bd-ab21-4286-b304-2acbab4c3ccf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0NjM0MTgsIm5iZiI6MTczNjQ2MzExOCwicGF0aCI6Ii82OTIwODcyNy80MDE3MTU3NTgtODI1MGY0YmQtYWIyMS00Mjg2LWIzMDQtMmFjYmFiNGMzY2NmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDIyNTE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExZDQyMmM2YmY5ZWIwZTQ5NGE3OTA5MzM1NWFiN2RjNzk3OWI0NTRkOWMyYmUxNTU3MjRjNWNiZDY1NTM2YTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h_yAoHZuf-AfwrikOIB2sosetGo8EMGDzfcl5z1ghKs)

You need to our LLM message semantic conventions. These can be a bit tricky due to constraints on OTel attribute value types. You can see an example [here](https://github.com/Arize-ai/openinference/blob/41ea8361508e00d34c7a23f9d3cdbdad56d0302b/python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py#L663).

axiomofjoy commented:
> @axiomofjoy one more problem that I just detected: detecting calls to [Model.get_tool_call](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/models.py#L160) from within [ToolCallingAgent](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/agents.py#L779) does not work. I think this is because I actually use [HfApiModel](https://github.com/huggingface/smolagents/blob/cb9830a5540d47cf624c359a3e2a37d8276cc42e/src/smolagents/models.py#L274) as my LLM: `HfApiModel` inherits from `Model` but since it overrides its parent's `get_tool_call` method, maybe the wrapper targeting `Model.get_tool_call()` does not work for `HfApiModel`.
> 
> Is there a way to create a wrapper that works for `Subclass.get_tool_call` for any subclass of `Model`?
> 
> Method `__call__` works for standard LLM calls (LLM calls that are just "please generate some text", not "please generate a tool call") because it's defined in Model and not overridden. But it would seem like a hack to build such a wrapper method for `Model.get_tool_call`, so I'd prefer the universal wrapper described above.
> 
> ![image](https://private-user-images.githubusercontent.com/69208727/401712782-4acccc83-fdb5-40c2-808e-beaca316a6c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0NjM0MTgsIm5iZiI6MTczNjQ2MzExOCwicGF0aCI6Ii82OTIwODcyNy80MDE3MTI3ODItNGFjY2NjODMtZmRiNS00MGMyLTgwOGUtYmVhY2EzMTZhNmM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDIyNTE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiZTg1YTdkZWE0ZmI5M2VmOTVhZGM5YTY4NjAwODJkMzA0ZTAyOGU2OGY1NTFiOWJlOWFkZmQ2ZGMyNDYxNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SuuGRQr6T7CSs6IHEhoV5PhIO_5fXujjB0S8jSVimMM)

Good catch! My first thought is that you might try instrumenting each subclass individually in addition to the base class. This can be accomplished by iterating over subclasses. We do something similar in DSPy [here](https://github.com/Arize-ai/openinference/blob/41ea8361508e00d34c7a23f9d3cdbdad56d0302b/python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py#L87).

axiomofjoy commented:
@aymeric-roucher It's looking really promising so far! A few findings from my initial testing.

1. A few span kinds are missing inputs and outputs (agent spans are missing input, LLM spans are missing both). We typically add these values for all OpenInference span kinds to show in the spans and traces table.

<img width="1555" alt="Screenshot 2025-01-09 at 4 27 56 PM" src="https://github.com/user-attachments/assets/bbce0b7c-bf7d-4b99-8776-161ec62ac02d" />

2. Tool attributes are missing. These attributes are the `tools` conventions in [this table](https://github.com/Arize-ai/openinference/blob/main/spec/semantic_conventions.md).

<img width="1241" alt="Screenshot 2025-01-09 at 4 29 09 PM" src="https://github.com/user-attachments/assets/e310900f-6ea1-40e1-a686-9ccd5c0b6432" />

3. Retriever tools might best be instrumented as retriever span kind rather than a tool to give us rich UI for retrieved documents, scores, etc.

If you don't mind, I'll open a PR against your branch including some of the examples I used!

aymeric-roucher commented:
![image](https://github.com/user-attachments/assets/32b8fcb9-c6d5-4193-92ff-51ee05bc8185)

I've applied your suggestions @axiomofjoy ! Here's what my dashboard now looks like.
Addressing your points:
1. Should be solved (cf screenshot)
2. I think tool attributes are now input correctly. Is there any attribute you still see missing?
3. We have no specific Retriever tool (or the community might build some but we can't know in advance how they'll work, they could return only a big string report instead of individual docs), so this has no use here IMO!

axiomofjoy commented:
Hey @aymeric-roucher, awesome progress! I'm excited to test it out.

I opened a PR to your fork that sets up CI and adds examples [here](https://github.com/aymeric-roucher/openinference/pull/1). This enables the following commands:

- `tox run -e mypy-smolagents` (run type checks)
- `tox run -e ruff-smolagents` (run formatters and linters)
- `tox run -e test-smolagents` (run tests, my PR nixes the CrewAI tests since they are out of date with our current patterns)
- `tox run -e smolagents` (run everything)

You can read about how to get set up with `tox` [here](https://github.com/Arize-ai/openinference/blob/main/python/DEVELOPMENT.md#development).

I'll start dogfooding your changes and add some tests in a subsequent PR if you don't mind!

aymeric-roucher commented:
Thanks a lot for your proposed changes @axiomofjoy! I'm so looking forward to enabling observability 🌟

axiomofjoy commented:
Began adding tests [here](https://github.com/aymeric-roucher/openinference/pull/2) @aymeric-roucher!

aymeric-roucher commented on python/instrumentation/openinference-instrumentation-smolagents/src/openinference/instrumentation/smolagents/_wrappers.py:
```diff
@@ -0,0 +1,442 @@
+                span.set_status(trace_api.Status(trace_api.StatusCode.ERROR, str(exception)))
+                span.record_exception(exception)
+                raise
+            span.set_attribute("Tool calls", step_log.tool_calls)
```
@axiomofjoy here's where Ive logged the tool calls (it's a list of ToolCall object, ToolCall is a dataclass so it could just be json-dumped)

aymeric-roucher commented:
Thank you for these changes @axiomofjoy !

Two questions from my side:

1. I've added logging for step_log.tool_calls in _StepWrapper
Do you know of a Chain attribute to log this list of tool calls (each tool call is a dataclass but they could be JSON dumped) in the Info tab above any error or output?

2. Is there a way to log errors in a less error-y way?
Below is an example where I've introduced a deliberate error, but then the agent recovers and successfully solves the task.
<img width="1132" alt="image" src="https://github.com/user-attachments/assets/693276c8-79b2-422c-a1ee-33a551347512" />
In `smolagents`, having an error in one step is not that much of an issue, it's normal for the LLM to give incorrect tool calls then recover in later steps.

So I want to show errors as part of the normal workflow. (else, users open issues like "my agent fails" and think it's a workflow issue instead of seeing that it's normal)

Could there be in a less red-flash-bad-error way to log errors? Like, not a frightening status error, but just a red-background message in the info tab, similar to how an Input or Output would be logged? Having the red (!) element on the minified representation is fine though.

aymeric-roucher commented:
But with these comments we're really entering the "small nits" realm, we can merge when you're ready!

axiomofjoy commented:
> Thank you for these changes @axiomofjoy !
> 
> Two questions from my side:
> 
> 1. I've added logging for step_log.tool_calls in _StepWrapper
>    Do you know of a Chain attribute to log this list of tool calls (each tool call is a dataclass but they could be JSON dumped) in the Info tab above any error or output?
> 2. Is there a way to log errors in a less error-y way?
>    Below is an example where I've introduced a deliberate error, but then the agent recovers and successfully solves the task.
> 
> <img alt="image" width="1132" src="https://private-user-images.githubusercontent.com/69208727/402511903-693276c8-79b2-422c-a1ee-33a551347512.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTk5NDgsIm5iZiI6MTczNjc5OTY0OCwicGF0aCI6Ii82OTIwODcyNy80MDI1MTE5MDMtNjkzMjc2YzgtNzliMi00MjJjLWExZWUtMzNhNTUxMzQ3NTEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDIwMjA0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2MjFlZTQ3OWMxMWEwZDUyN2Q1MzdiYzJkNWU0MTk2YWMyYTM4NDZkNGU1ZTY2NjM4Zjc1YWVmMWRmYTZjYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mue8VClWKNvJu0MLNeJwyTqJAyJgjHy49D0WFkMIuAA"> In `smolagents`, having an error in one step is not that much of an issue, it's normal for the LLM to give incorrect tool calls then recover in later steps.
> So I want to show errors as part of the normal workflow. (else, users open issues like "my agent fails" and think it's a workflow issue instead of seeing that it's normal)
> 
> Could there be in a less red-flash-bad-error way to log errors? Like, not a frightening status error, but just a red-background message in the info tab, similar to how an Input or Output would be logged? Having the red (!) element on the minified representation is fine though.

Hey @aymeric-roucher! We don't support a mechanism to change the salience of error messages in the UI. If you want to log information that is similar to an error, another option would be to log an OTel event without setting an error status code. This would result in a message in the events tab, but no (!).

aymeric-roucher commented:
Ok, thank you for the precision. Then all's good on my side! Do you want help to make sure all tests pass/ add additional tests?

axiomofjoy commented:
One more PR [here](https://github.com/aymeric-roucher/openinference/pull/3) with additional tests and tweaks to tool spans, CI, and `smolagents`-specific attributes. This should get us passing CI as well. If the telemetry still looks good in the UI to you @aymeric-roucher, I think we're ready for an initial release!

> Ok, thank you for the precision. Then all's good on my side! Do you want any help, for instance to make sure all tests pass/ add any more required tests?

It would be helpful for maintainability if we had a test case with a minimalistic agent that produces relatively consistent output that would give us coverage on the `_RunWrapper` and `_StepWrapper`. This is not blocking the initial release, though 😁

aymeric-roucher commented:
@axiomofjoy just added an agent that do just this! I've only put smolagent-side tests for now, not added the OTel-side testing (I figured you would be faster than me to make these tests), but I can do it if you want!

aymeric-roucher commented:
Also @axiomofjoy do we have way to change all errors into warnings ? This may solve the scary aspect of them, showing that errors are a normal part of an agent's run.

axiomofjoy commented:
> @axiomofjoy just added tests with deterministic agent objects! I've only put smolagent-side tests for now, not added the OTel-side testing (I figured you would be faster than me to make these tests), but I can do it if you want!

Thanks so much @aymeric-roucher! We can add in a follow-up.

I have one more PR still open [here](https://github.com/aymeric-roucher/openinference/pull/3) if you can take a look!

aymeric-roucher commented:
It's merged @axiomofjoy !

axiomofjoy commented:
Amazing work @aymeric-roucher! So excited to get this merged.

I've filed an umbrella ticket [here](https://github.com/Arize-ai/openinference/issues/1195) to track fast follows, including setting up the release pipeline and addressing some of the outstanding issues in this thread. Please feel free to add to the issue!

We'll get an initial version of the package released tomorrow! 🎉

axiomofjoy commented:
Initial release is out [here](https://pypi.org/project/openinference-instrumentation-smolagents/)!

axiomofjoy commented:
Noticed a few small bugs and filed issues in the ticket. @aymeric-roucher please let me know how it looks to you! In particular, I have not yet been able to test with the `HfApiModel` since I am having some trouble getting set up with Inferences. Definitely want to make sure that is working.

aymeric-roucher commented:
@axiomofjoy I've updated our internal ChatMessages format to give it a `model_dumps_json()` method, now it works!
I've now released the telemetry in this release of smolagents: https://github.com/huggingface/smolagents/releases/tag/v1.3.0

axiomofjoy commented:
> @axiomofjoy I've updated our internal ChatMessages format to give it a `model_dumps_json()` method, now it works! I've now released the telemetry in this release of smolagents: https://github.com/huggingface/smolagents/releases/tag/v1.3.0

Amazing, thanks so much @aymeric-roucher!
